File: app/src/main/java/maderski/bluetoothautoplaymusic/BluetoothActions/BTConnectActions.java
package maderski.bluetoothautoplaymusic.BluetoothActions;

import android.app.KeyguardManager;
import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.IntentFilter;
import android.os.Build;
import android.os.CountDownTimer;
import android.os.Handler;
import android.util.Log;

import maderski.bluetoothautoplaymusic.Controls.PlayMusicControl;
import maderski.bluetoothautoplaymusic.Controls.RingerControl;
import maderski.bluetoothautoplaymusic.Controls.VolumeControl;
import maderski.bluetoothautoplaymusic.Controls.WifiControl;
import maderski.bluetoothautoplaymusic.Helpers.PermissionHelper;
import maderski.bluetoothautoplaymusic.Helpers.PowerHelper;
import maderski.bluetoothautoplaymusic.Helpers.TimeHelper;
import maderski.bluetoothautoplaymusic.Helpers.LaunchAppHelper;
import maderski.bluetoothautoplaymusic.Notification;
import maderski.bluetoothautoplaymusic.Receivers.NotifPolicyAccessChangedReceiver;
import maderski.bluetoothautoplaymusic.Services.OnBTConnectService;
import maderski.bluetoothautoplaymusic.Services.WakeLockService;
import maderski.bluetoothautoplaymusic.SharedPrefs.BAPMDataPreferences;
import maderski.bluetoothautoplaymusic.SharedPrefs.BAPMPreferences;
import maderski.bluetoothautoplaymusic.Telephone;
import maderski.bluetoothautoplaymusic.Utils.ServiceUtils;

/**
 * Created by Jason on 6/3/17.
 */

public class BTConnectActions {
    private static final String TAG = "BTConnectActions";

    private final Context context;
    private final Notification mNotification;
    private final VolumeControl mVolumeControl;
    private final PlayMusicControl mPlayMusicControl;
    private final LaunchAppHelper mLaunchAppHelper;

    public BTConnectActions(Context context){
        this.context = context;
        mNotification = new Notification();
        mVolumeControl = new VolumeControl(context);
        mPlayMusicControl = new PlayMusicControl(context);
        mLaunchAppHelper = new LaunchAppHelper();
    }

    public void OnBTConnect(){
        boolean waitTillOffPhone = BAPMPreferences.getWaitTillOffPhone(context);

        if(waitTillOffPhone){
            Telephone telephone = new Telephone(context);
            if(PowerHelper.isPluggedIn(context)){
                if(telephone.isOnCall()) {
                    Log.d(TAG, "ON a call");
                    //Run CheckIfOnPhone
                    telephone.CheckIfOnPhone(mVolumeControl);
                }else{
                    Log.d(TAG, "NOT on a call");
                    actionsOnBTConnect();
                }
            }else{
                if(telephone.isOnCall()) {
                    mNotification.launchBAPM(context);
                }else{
                    actionsOnBTConnect();
                }
            }
        }else{
            actionsOnBTConnect();
        }
    }

    //Creates mNotification and if set turns screen ON, puts the phone in priority mode,
    //sets the volume to MAX, dismisses the keyguard, Launches the Music Selected Music
    //Player and Launches Maps
    public void actionsOnBTConnect(){
        final boolean unlockScreen = BAPMPreferences.getUnlockScreen(context);

        showBTAMNotification();
        setVolumeToMax();
        turnTheScreenOn();

        if(unlockScreen){
            performActionsDelay();
        } else {
            launchMusicMapApp();
            autoPlayMusic(6);
            setWifiOff();
            putPhoneInDoNotDisturb();
        }

        BAPMDataPreferences.setRanActionsOnBtConnect(context, true);
        ServiceUtils.stopService(context, OnBTConnectService.class, OnBTConnectService.TAG);
    }

    private void performActionsDelay(){
        int seconds = 30000;

        new CountDownTimer(seconds, 1000){

            @Override
            public void onTick(long millisUntilFinished) {
                KeyguardManager keyguardManager = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
                boolean isDeviceLocked = keyguardManager.isDeviceLocked();
                Log.d(TAG, "IS DEVICE LOCKED: " + String.valueOf(isDeviceLocked));
                if (!isDeviceLocked && millisUntilFinished < 28000) {
                    cancel();
                    onFinish();
                }
                Log.d(TAG, "LOCKED mills left to check: " + String.valueOf(millisUntilFinished));
            }

            @Override
            public void onFinish() {
                unlockTheScreen();
                launchMusicMapApp();
                autoPlayMusic(6);
                setWifiOff();
                putPhoneInDoNotDisturb();
            }
        }.start();
    }

    private void showBTAMNotification(){
        String mapChoice = BAPMPreferences.getMapsChoice(context);
        boolean canShowNotification = BAPMPreferences.getShowNotification(context);
        if (canShowNotification) {
            mNotification.BAPMMessage(context, mapChoice);
        }
    }

    private void turnTheScreenOn(){
        boolean screenON = BAPMPreferences.getKeepScreenON(context);
        if (screenON) {
            ServiceUtils.startService(context, WakeLockService.class, WakeLockService.TAG);
        }
    }

    private void unlockTheScreen(){
        KeyguardManager keyguardManager = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
        boolean isKeyguardLocked = keyguardManager.isKeyguardLocked();
        Log.d(TAG, "Is keyguard locked: " + Boolean.toString(isKeyguardLocked));
        if (isKeyguardLocked) {
            mLaunchAppHelper.launchBAPMActivity(context);
        }
    }

    private void setVolumeToMax(){
        boolean volumeMAX = BAPMPreferences.getMaxVolume(context);
        if (volumeMAX) {
            Handler handler = new Handler();
            Runnable runnable = new Runnable() {
                @Override
                public void run() {
                    mVolumeControl.checkSetMAXVol(4);
                }
            };
            handler.postDelayed(runnable, 3000);
        }
    }

    private void autoPlayMusic(final int checkToPlaySeconds){
        boolean playMusic = BAPMPreferences.getAutoPlayMusic(context);
        if (playMusic) {
            mPlayMusicControl.checkIfPlaying(context, checkToPlaySeconds);
        }
    }

    private void launchMusicMapApp(){
        boolean launchMusicPlayer = BAPMPreferences.getLaunchMusicPlayer(context);
        boolean launchMaps = BAPMPreferences.getLaunchGoogleMaps(context);
        boolean mapsCanLaunch = mLaunchAppHelper.canMapsLaunchNow(context);

        if (launchMusicPlayer && !launchMaps || launchMusicPlayer && !mapsCanLaunch) {
            mLaunchAppHelper.musicPlayerLaunch(context, 3);
        }

        if (launchMaps) {
            mLaunchAppHelper.launchMaps(context, 3);
        }
    }

    private void setWifiOff(){
        boolean isWifiOffDevice = BAPMDataPreferences.getIsTurnOffWifiDevice(context);
        if (isWifiOffDevice) {
            int morningStartTime = BAPMPreferences.getMorningStartTime(context);
            int morningEndTime = BAPMPreferences.getMorningEndTime(context);

            int current24hrTime = TimeHelper.getCurrent24hrTime();

            TimeHelper timeHelperMorning = new TimeHelper(morningStartTime, morningEndTime, current24hrTime);
            boolean canLaunch = timeHelperMorning.isWithinTimeSpan();
            String directionLocation = canLaunch ? LaunchAppHelper.DirectionLocations.WORK : LaunchAppHelper.DirectionLocations.HOME;

            boolean canChangeWifiState = !BAPMPreferences.getWifiUseMapTimeSpans(context)
                    || (canLaunch && mLaunchAppHelper.canLaunchOnThisDay(context, directionLocation));
            if (canChangeWifiState && WifiControl.isWifiON(context)) {
                WifiControl.wifiON(context, false);
            }
        }
    }

    private void putPhoneInDoNotDisturb(){
        RingerControl ringerControl = new RingerControl(context);
        boolean priorityMode = BAPMPreferences.getPriorityMode(context);

        if (priorityMode) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                boolean hasDoNotDisturbPerm = PermissionHelper.checkDoNotDisturbPermission(context, 10);
                if (hasDoNotDisturbPerm) {
                    BAPMDataPreferences.setCurrentRingerSet(context, ringerControl.ringerSetting());
                    ringerControl.soundsOFF();
                } else {
                    BroadcastReceiver broadcastReceiver = new NotifPolicyAccessChangedReceiver();
                    IntentFilter intentFilter = new IntentFilter(NotificationManager.ACTION_NOTIFICATION_POLICY_ACCESS_GRANTED_CHANGED);
                    context.getApplicationContext().registerReceiver(broadcastReceiver, intentFilter);
                }
            } else {
                BAPMDataPreferences.setCurrentRingerSet(context, ringerControl.ringerSetting());
                ringerControl.soundsOFF();
            }
        }
    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/bluetoothactions/BTConnectActions.kt
package maderski.bluetoothautoplaymusic.bluetoothactions

import android.app.KeyguardManager
import android.app.NotificationManager
import android.content.Context
import android.content.IntentFilter
import android.os.Build
import android.os.CountDownTimer
import android.os.Handler
import android.os.Looper
import android.util.Log
import maderski.bluetoothautoplaymusic.controls.RingerControl
import maderski.bluetoothautoplaymusic.controls.VolumeControl
import maderski.bluetoothautoplaymusic.controls.WifiControl
import maderski.bluetoothautoplaymusic.controls.mediaplayer.MediaPlayerControlManager
import maderski.bluetoothautoplaymusic.helpers.LaunchAppHelper
import maderski.bluetoothautoplaymusic.helpers.LaunchAppHelper.DirectionLocation
import maderski.bluetoothautoplaymusic.helpers.TelephoneHelper
import maderski.bluetoothautoplaymusic.helpers.TimeHelper
import maderski.bluetoothautoplaymusic.notification.BAPMNotification
import maderski.bluetoothautoplaymusic.receivers.NotifPolicyAccessChangedReceiver
import maderski.bluetoothautoplaymusic.services.OnBTConnectService
import maderski.bluetoothautoplaymusic.services.ServiceManager
import maderski.bluetoothautoplaymusic.services.WakeLockService
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMDataPreferences
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences
import maderski.bluetoothautoplaymusic.utils.PermissionUtils
import maderski.bluetoothautoplaymusic.helpers.PowerHelper
import org.koin.core.KoinComponent
import org.koin.core.inject

/**
 * Created by Jason on 6/3/17.
 */

class BTConnectActions(private val context: Context): KoinComponent {
    private val preferences: BAPMPreferences by inject()
    private val dataPreferences: BAPMDataPreferences by inject()
    private val volumeControl: VolumeControl by inject()
    private val bapmNotification: BAPMNotification by inject()
    private val launchAppHelper: LaunchAppHelper by inject()
    private val mediaPlayerControlManager: MediaPlayerControlManager by inject()
    private val serviceManager: ServiceManager by inject()
    private val powerHelper: PowerHelper by inject()
    private val telephoneHelper: TelephoneHelper by inject()

    fun onBTConnect() {
        val waitTillOffPhone = preferences.getWaitTillOffPhone()
        if (waitTillOffPhone) {
            actionsWhileOnCall()
        } else {
            actionsOnBTConnect()
        }
    }

    private fun actionsWhileOnCall() {
        val isOnCall = telephoneHelper.isOnCall
        val isPluggedIn = powerHelper.isPluggedIn()

        if (isOnCall) {
            Log.d(TAG, "ON a call")
            // if plugged in run on the phone check else launch notification
            if (isPluggedIn) telephoneHelper.checkIfOnPhone(volumeControl) else bapmNotification.launchBAPM()
        } else {
            Log.d(TAG, "NOT on a call")
            actionsOnBTConnect()
        }
    }

    //Creates mNotification and if set turns screen ON, puts the phone in priority mode,
    //sets the volume to MAX, dismisses the keyguard, Launches the Music Selected Music
    //Player and Launches Maps
    fun actionsOnBTConnect() {
        val unlockScreen = preferences.getUnlockScreen()

        showBTAMNotification()
        setVolumeToMax()
        turnTheScreenOn()

        if (unlockScreen) {
            performActionsDelay()
        } else {
            launchMusicMapApp()
            autoPlayMusic()
            setWifiOff()
            putPhoneInDoNotDisturb()
        }

        serviceManager.stopService(OnBTConnectService::class.java, OnBTConnectService.TAG)
    }

    private fun performActionsDelay() {
        Handler(Looper.getMainLooper()).post {
            val milliSeconds = 30000
            object : CountDownTimer(milliSeconds.toLong(), 1000) {

                override fun onTick(millisUntilFinished: Long) {
                    val keyguardManager = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
                    val isDeviceLocked = keyguardManager.isDeviceLocked
                    Log.d(TAG, "IS DEVICE LOCKED: $isDeviceLocked")
                    if (!isDeviceLocked && millisUntilFinished < 28000) {
                        cancel()
                        onFinish()
                    }
                    Log.d(TAG, "LOCKED mills left to check: $millisUntilFinished")
                }

                override fun onFinish() {
                    unlockTheScreen()
                    launchMusicMapApp()
                    autoPlayMusic()
                    setWifiOff()
                    putPhoneInDoNotDisturb()
                }
            }.start()
        }
    }

    private fun showBTAMNotification() {
        val mapChoice = preferences.getMapsChoice()
        val canShowNotification = preferences.getShowNotification()
        if (canShowNotification) {
            bapmNotification.bapmMessage(mapChoice)
        }
    }

    private fun turnTheScreenOn() {
        val screenON = preferences.getKeepScreenON()
        if (screenON) {
            serviceManager.startService(WakeLockService::class.java, WakeLockService.TAG)
        }
    }

    private fun unlockTheScreen() {
        val keyguardManager = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
        val isKeyguardLocked = keyguardManager.isKeyguardLocked
        Log.d(TAG, "Is keyguard locked: " + java.lang.Boolean.toString(isKeyguardLocked))
        if (isKeyguardLocked) {
            launchAppHelper.launchBAPMActivity()
        }
    }

    private fun setVolumeToMax() {
        val volumeMAX = preferences.getMaxVolume()
        if (volumeMAX) {
            volumeControl.checkSetMAXVol(4)
        }
    }

    private fun autoPlayMusic() {
        val playMusic = preferences.getAutoPlayMusic()
        if (playMusic) {
            mediaPlayerControlManager.play()
        }
    }

    private fun launchMusicMapApp() {
        val launchMusicPlayer = preferences.getLaunchMusicPlayer()
        val launchMaps = preferences.getLaunchGoogleMaps()
        val mapsCanLaunch = launchAppHelper.canMapsLaunchNow()

        if (launchMusicPlayer && !launchMaps || launchMusicPlayer && !mapsCanLaunch) {
            launchAppHelper.musicPlayerLaunch(3)
        }

        if (launchMaps) {
            launchAppHelper.launchMaps(3)
        }
    }

    private fun setWifiOff() {
        val isWifiOffDevice = dataPreferences.getIsTurnOffWifiDevice()
        if (isWifiOffDevice) {
            val morningStartTime = preferences.getMorningStartTime()
            val morningEndTime = preferences.getMorningEndTime()

            val current24hrTime = TimeHelper.current24hrTime

            val timeHelperMorning = TimeHelper(morningStartTime, morningEndTime, current24hrTime)
            val canLaunch = timeHelperMorning.isWithinTimeSpan
            val directionLocation = if (canLaunch) DirectionLocation.WORK else DirectionLocation.HOME

            val canChangeWifiState = !preferences.getWifiUseMapTimeSpans() || canLaunch && launchAppHelper.canLaunchOnThisDay(directionLocation)
            if (canChangeWifiState && WifiControl.isWifiON(context)) {
                WifiControl.wifiON(context, false)
            }
        }
    }

    private fun putPhoneInDoNotDisturb() {
        val ringerControl = RingerControl(context)
        val priorityMode = preferences.getPriorityMode()

        if (priorityMode) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                val hasDoNotDisturbPerm = PermissionUtils.checkDoNotDisturbPermission(context, 10)
                if (hasDoNotDisturbPerm) {
                    dataPreferences.setCurrentRingerSet(ringerControl.ringerSetting())
                    ringerControl.soundsOFF()
                } else {
                    val broadcastReceiver = NotifPolicyAccessChangedReceiver()
                    val intentFilter = IntentFilter(NotificationManager.ACTION_NOTIFICATION_POLICY_ACCESS_GRANTED_CHANGED)
                    context.applicationContext.registerReceiver(broadcastReceiver, intentFilter)
                }
            } else {
                dataPreferences.setCurrentRingerSet(ringerControl.ringerSetting())
                ringerControl.soundsOFF()
            }
        }
    }

    companion object {
        private const val TAG = "BTConnectActions"
    }
}
